src/com/ka/spreadsheet/diff/UnifiedDiffSpreadSheetDiffCallback.java

```diff
@@ -0,0 +1,172 @@
+
+  // This formatter requires that each cell in the pair of workbooks will be processed in worksheet-row-column order.
+  // SpreadSheetDiffer.java does that, but it's not a contract.  If that changes some day, set this to true to resort
+  // the list of diffs/extras before formatting them.
+  private static final boolean diffCellsIsUnsorted = false;
+
+  @Override
+  public void reportWorkbooksDiffer(boolean differ, File file1, File file2) {
+    String sheetName = "";
+    int row = -1;
+    int col = -1;
+    List<DiffCell> cellBlock = null;
+    sortList(diffCells);
+    for (DiffCell diffCell : diffCells) {
+     if (!diffCell.sheetName.equals(sheetName)) {
+        // New sheet, finish the active row (if any) and start the new sheet.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        sheetName = diffCell.sheetName;
+        System.out.println("--- " + file1 + "!" + sheetName);
+        System.out.println("+++ " + file2 + "!" + sheetName);
+        cellBlock = null;
+        row = -1;
+      }
+      if (diffCell.rowIndex != row) {
+        // New row, finish the active row (if any) and start the new row.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        cellBlock = null;
+        row = diffCell.rowIndex;
+        col = -1;
+      }
+      if (diffCell.colIndex != col) {
+        // New cell-block, finish the active cell-block (if any) and start a new cell-block.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        cellBlock = new ArrayList<DiffCell>();
+      }
+      // Add this cell to the active cell-block.
+      cellBlock.add(diffCell);
+      col = diffCell.colIndex + 1;
+    }
+    if (cellBlock != null) {
+      printRow(cellBlock);
+    }
+  }
+
+  @Override
+  public void reportMacroOnlyIn(boolean inFirstSpreadSheet) {
+    System.out.println("Unified diff format does not support macros, however WB" + (inFirstSpreadSheet ? "1" : "2") + " contains at least one macro that is not in the other workbook.");
+  }
+
+  @Override
+  public void reportExtraCell(boolean inFirstSpreadSheet, CellPos c) {
+    diffCells.add(new DiffCell(
+      c.getSheetName(),
+      c.getRowIndex(),
+      c.getColumnIndex(),
+      (inFirstSpreadSheet ? c.getCellValue() : null),
+      (!inFirstSpreadSheet ? c.getCellValue() : null)
+    ));
+  }
+
+  @Override
+  public void reportDiffCell(CellPos c1, CellPos c2) {
+    if (c1.getRowIndex() != c1.getRowIndex() || c2.getColumnIndex() != c2.getColumnIndex()) {
+        throw new RuntimeException("Invalid cell comparison: WB1=" + c1.getCellPosition() + ", WB2=" + c2.getCellPosition());
+    }
+    diffCells.add(new DiffCell(
+      c1.getSheetName(),
+      c1.getRowIndex(),
+      c1.getColumnIndex(),
+      c1.getCellValue(),
+      c2.getCellValue()
+    ));
+  }
+
+  // TODO: Make this handle multiple rows, maybe.  What would that output look like?  This might be a bad idea.
```

Comment:
This is the hint about rectangular contiguity.  Doing this might actually appear uglier than not doing it.

Reply:
Right